kv: de-flake TestRangeCacheDetectSplitReverseScan

tbg opened the pull request:
The test was occasionally failing with only three out of four expected lookups
in the [a,az) case because certain interleavings of the two parallel lookups
could populate each other's cache, sometimes avoiding one lookup.

Verifying that the number of lookups in that case is either three or four
(instead of requiring four) solves that issue.

Fixes #6783.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6884)

<!-- Reviewable:end -->

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> #### kv: de\-flake TestRangeCacheDetectSplitReverseScan
> 
> The test was occasionally failing with only three out of four expected lookups
> in the [a,az) case because certain interleavings of the two parallel lookups
> could populate each other's cache, sometimes avoiding one lookup.
> 
> Verifying that the number of lookups in that case is either three or four
> (instead of requiring four) solves that issue.
> 
> Fixes #6783.
> 
> 
> 
</details>


---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[kv/range_cache_test.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6884#-KIZkiPfz1ZyuVNR1ATb:-KIZkiPfz1ZyuVNR1ATc:-200605787) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b3608cf594f9dfa9de4b528ea97ad23a392e02/kv/range_cache_test.go#L188)):*
> ```Go
> }
> 
> func (db *testDescriptorDB) assertLookupCount(t *testing.T, from, to int64, key string) {
> ```

Can you add a quick comments on assertLookupCount and asserLookupCountEq.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6884#-:-KIZlF08JFiVsBfv2iSF:1417088721)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, although I wish we could avoid a workaround for nondeterminism in tests. Is the issue you are seeing caused by descriptor prefetching?

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6884#-:-KIa3k59ooDwgymo2UXH:-411632576)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'm not sure. I tried disabling prefetching, but it throws off assertions earlier in the test. I agree with you in principle but don't want to spend more time on it.

tbg committed:
kv: de-flake TestRangeCacheDetectSplitReverseScan

The test was occasionally failing with only three out of four expected lookups
in the [a,az) case because certain interleavings of the two parallel lookups
could populate each other's cache, sometimes avoiding one lookup.

Verifying that the number of lookups in that case is either three or four
(instead of requiring four) solves that issue.

Fixes #6783.

tschottdorf commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[kv/range_cache_test.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6884#-KIZkiPfz1ZyuVNR1ATb:-KId4WfDtiw3L86KAmG-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b3608cf594f9dfa9de4b528ea97ad23a392e02/kv/range_cache_test.go#L188)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Can you add a quick comments on assertLookupCount and asserLookupCountEq.  
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6884)*
<!-- Sent from Reviewable.io -->
